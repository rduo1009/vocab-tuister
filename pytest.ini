[pytest]
addopts = 
    --strict-markers
    --ignore=src/_vendor
    -n4

markers =
    manual: marks tests as manually checked