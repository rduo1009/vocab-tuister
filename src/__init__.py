"""A wrapper for the Python packages used by ``vocab-tester``."""

from __future__ import annotations

import os as _os

_seed: str | None = _os.getenv("VOCAB_TUISTER_RANDOM_SEED")
if _seed is not None:
    import random
    import sys as _sys
    import types as _types

    try:
        _seed_value: int = int(_seed)
    except ValueError as e:
        raise ValueError(
            f"Invalid seed value: {_seed}. Must be an integer."
        ) from e

    _custom_random = _types.ModuleType("random")
    _custom_random_class = random.Random(_seed_value)

    _custom_random.Random = random.Random  # type: ignore[attr-defined]
    _custom_random.SystemRandom = random.SystemRandom  # type: ignore[attr-defined]

    _custom_random.random = _custom_random_class.random  # type: ignore[attr-defined]
    _custom_random.uniform = _custom_random_class.uniform  # type: ignore[attr-defined]

    _custom_random.randint = _custom_random_class.randint  # type: ignore[attr-defined]
    _custom_random.randrange = _custom_random_class.randrange  # type: ignore[attr-defined]

    _custom_random.choice = _custom_random_class.choice  # type: ignore[attr-defined]
    _custom_random.sample = _custom_random_class.sample  # type: ignore[attr-defined]
    _custom_random.shuffle = _custom_random_class.shuffle  # type: ignore[attr-defined]

    _sys.modules["random"] = _custom_random


import dunamai as _dunamai

from . import core, server

try:
    __version__ = _dunamai.get_version(
        "src", third_choice=_dunamai.Version.from_any_vcs
    ).serialize()
except RuntimeError:
    import sys as _sys

    # Frozen with PyInstaller
    if getattr(_sys, "frozen", False) and hasattr(
        _sys, "_MEIPASS"
    ):  # pragma: no cover
        from pathlib import Path

        version_path = Path(__file__).parent.parent / "__version__.txt"

        __version__ = version_path.read_text().strip()
    else:
        raise

__author__ = "rduo1009"
__copyright__ = "2024, rduo1009"
__license__ = "MIT"
__all__ = ["core", "server"]
