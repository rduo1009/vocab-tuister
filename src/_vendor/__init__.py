from . import aenum, typeddict_validator