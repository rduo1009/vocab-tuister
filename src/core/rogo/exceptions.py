"""Contains custom exceptions used by ``rogo``."""


class InvalidSessionConfigError(Exception):
    """An error that is raised when the given session config is invalid."""
