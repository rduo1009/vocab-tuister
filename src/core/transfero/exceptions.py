"""Contains custom exceptions used by ``transfero``."""


class InvalidWordError(Exception):
    """An error that is raised when the word is invalid."""


class InvalidComponentsError(Exception):
    """An error that is raised when the components are invalid."""
