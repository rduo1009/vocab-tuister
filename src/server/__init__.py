"""Contains the server for vocab-tester."""

from . import app, exceptions

__all__ = ["app", "exceptions"]
