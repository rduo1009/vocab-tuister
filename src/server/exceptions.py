"""Contains custom exceptions used by the server."""


class InvalidSettingsError(Exception):
    """An error that is raised when the given settings are invalid."""
