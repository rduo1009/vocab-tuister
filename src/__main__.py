"""The CLI that runs the server."""

# ruff: noqa: TCH002

from __future__ import annotations

import logging
import sys
import warnings
from typing import Annotated

from cyclopts import App, Parameter
from cyclopts.types import UInt16

from src import __version__
from src.utils.logger import (
    CustomHandler,
    custom_formatwarning,
    log_uncaught_exceptions,
)

from .server.app import main as server_main
from .server.app import main_dev as server_main_dev

# Initialise logger
logger = logging.getLogger()
logging.captureWarnings(capture=True)

# Set custom formatting for warnings and exceptions
warnings.formatwarning = custom_formatwarning
sys.excepthook = log_uncaught_exceptions  # type: ignore[assignment]

# Set custom handler for logs
for handler in logger.handlers.copy():
    logger.removeHandler(handler)
handler = CustomHandler()


logger.addHandler(handler)
cli = App(
    version=__version__,  # for some reason this is needed
    default_parameter=Parameter(negative=()),
)


def _set_verbosity(verbosity: int) -> None:
    verbosity_map = {
        -1: logging.CRITICAL,  # quiet mode
        0: logging.ERROR,
        1: logging.WARNING,
        2: logging.INFO,
        3: logging.DEBUG,
    }

    logger.setLevel(verbosity_map.get(verbosity, logging.DEBUG))
    handler.setLevel(verbosity_map.get(verbosity, logging.DEBUG))


@cli.default
def vocab_tuister_server(
    *,
    port: Annotated[UInt16, Parameter(name=["--port", "-p"])] = 5000,
    verbose: Annotated[
        tuple[bool, ...], Parameter(name=["--verbose", "-v"])
    ] = (False,),
    quiet: bool = False,
    dev: bool = False,
    debug: bool = False,
) -> None:
    """Start the vocab-tuister server.

    Parameters
    ----------
    port : int
        The port to run the server on.
    verbose : tuple[bool, ...]
        How verbose to make the output (maximum 3). Note that additional
        verbosity flags must be added like ``-v -v -v``, due to a bug with
        Cyclopts.
    quiet : bool
        Whether to make the output quiet (only critical errors). Overrides
        ``--verbose``.
    dev : bool
        Use the Flask development server instead of the production server.
        Should not be used usually.
    debug : bool
        Run in debug mode. Implies ``--dev``.
        Should not be used usually.
    """
    _set_verbosity(-1 if quiet else sum(verbose))

    if dev or debug:
        server_main_dev(port, debug=debug)
    else:
        server_main(port)


if __name__ == "__main__":
    cli()
