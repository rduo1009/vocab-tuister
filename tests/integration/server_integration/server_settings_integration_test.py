# pyright: basic
# pyright: reportUnknownParameterType=false, reportUnknownArgumentType=false, reportMissingParameterType=false

import contextlib
import threading
import time
from pathlib import Path

import pytest
from src.__main__ import cli
from src.server.exceptions import InvalidSettingsError


def run_cli(port):
    with contextlib.suppress(KeyboardInterrupt):
        cli(["-v", "-v", "-v", "-p", str(port)])


def setup_tests(monkeypatch, port, settings_info):
    class _Dirs:
        user_config_path = Path(__file__).parent / "testdata" / f"test-{settings_info}-settings"
        user_cache_path = Path(__file__).parent / "testdata" / "cache"

    monkeypatch.setattr("src.server.app.vocab_list", None)
    monkeypatch.setattr("src.server.app.dirs", _Dirs)

    run_cli(port)


@pytest.mark.integration
def test_cli_normal_settings(monkeypatch):
    thread = threading.Thread(target=lambda: setup_tests(monkeypatch, 5510, "regular"), daemon=True)
    thread.start()

    time.sleep(2)


@pytest.mark.integration
def test_cli_error_missing_settings(snapshot, monkeypatch):
    with pytest.raises(InvalidSettingsError) as error:
        setup_tests(monkeypatch, 5511, "errormissing")
    assert str(error.value) == snapshot


@pytest.mark.integration
def test_cli_error_extra_settings(snapshot, monkeypatch):
    with pytest.raises(InvalidSettingsError) as error:
        setup_tests(monkeypatch, 5512, "errorextra")
    assert str(error.value) == snapshot


@pytest.mark.integration
def test_cli_error_type_settings(snapshot, monkeypatch):
    with pytest.raises(InvalidSettingsError) as error:
        setup_tests(monkeypatch, 5513, "errortype")
    assert str(error.value) == snapshot
