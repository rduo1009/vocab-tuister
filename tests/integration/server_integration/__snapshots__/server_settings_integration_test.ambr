# serializer version: 1
# name: test_cli_error_extra_settings
  "Unrecognised settings were provided: 'cause-error'."
# ---
# name: test_cli_error_missing_settings
  "Required settings are missing: 'include-similar-words'."
# ---
# name: test_cli_error_type_settings
  "Expected type typing.ReadOnly[bool] for 'include-similar-words', but received type int."
# ---
